src/object_store/src/object/opendal_engine/opendal_object_store.rs

```diff
@@ -0,0 +1,365 @@
+    }
+
+    /// Returns a stream reading the object specified in `path`. If given, the stream starts at the
+    /// byte with index `start_pos` (0-based). As far as possible, the stream only loads the amount
+    /// of data into memory that is read from the stream.
+    async fn streaming_read(
+        &self,
+        path: &str,
+        start_pos: Option<usize>,
+    ) -> ObjectResult<Box<dyn AsyncRead + Unpin + Send + Sync>> {
+        fail_point!("opendal_streaming_read_err", |_| Err(
+            ObjectError::internal("opendal streaming read error")
+        ));
+
+        let reader = match start_pos {
+            Some(start_position) => {
+                self.op
+                    .object(path)
+                    .range_reader(start_position as u64..)
+                    .await?
+            }
+            None => self.op.object(path).reader().await?,
+        };
+
+        Ok(Box::new(reader))
+    }
+
+    async fn metadata(&self, path: &str) -> ObjectResult<ObjectMetadata> {
+        let opendal_metadata = self.op.object(path).metadata().await?;
+        let key = path.to_string();
+        let last_modified = match opendal_metadata.last_modified() {
+            Some(t) => t.unix_timestamp() as f64,
+            None => 0_f64,
+        };
+
+        let total_size = opendal_metadata.content_length() as usize;
+        let metadata = ObjectMetadata {
+            key,
+            last_modified,
+            total_size,
+        };
+        Ok(metadata)
+    }
+
+    async fn delete(&self, path: &str) -> ObjectResult<()> {
+        self.op.object(path).delete().await?;
+        Ok(())
+    }
+
+    /// Deletes the objects with the given paths permanently from the storage. If an object
+    /// specified in the request is not found, it will be considered as successfully deleted.
+    async fn delete_objects(&self, paths: &[String]) -> ObjectResult<()> {
+        for path in paths {
+            self.op.object(path).delete().await?;
+        }
+        Ok(())
+    }
+
+    async fn list(&self, prefix: &str) -> ObjectResult<Vec<ObjectMetadata>> {
+        let mut object_lister = self.op.object(prefix).list().await?;
+        let mut metadata_list = vec![];
+        while let Some(obj) = object_lister.next().await {
+            let object = obj?;
+            let key = prefix.to_string();
+            let last_modified = match object.metadata().await?.last_modified() {
+                Some(t) => t.unix_timestamp() as f64,
+                None => 0_f64,
+            };
+
+            let total_size = object.metadata().await?.content_length() as usize;
+            let metadata = ObjectMetadata {
+                key,
+                last_modified,
+                total_size,
+            };
+            metadata_list.push(metadata);
+        }
+        Ok(metadata_list)
+    }
+
+    fn store_media_type(&self) -> &'static str {
+        match self.engine_type {
+            EngineType::Memory => "Memory",
```

Comment:
OpenDAL provides [`opendal::Scheme`](https://docs.rs/opendal/0.25.0/opendal/enum.Scheme.html) that implements `Into<&'static str>`, maybe worth a try.

Reply:
Thank you very much for your suggestions and attention to RisingWave😄, I will resolve these issues ASAP. 
In the future, we hope to provide several non-s3 object storage through OpenDAL, and look forward to our follow-up cooperation🙌

Reply:
> In the future, we hope to provide several non-s3 object storage through OpenDAL, and look forward to our follow-up cooperation

Coooool! It's an honor to be used in RisingWave!